doc/developer/platform/formalism.md

```diff
@@ -194,3 +199,111 @@ A view introduces *constraints* on the capabilities and frontiers of its input a
+
+---
+
+## Sources
+
+Materialize supports a `CREATE SOURCE` command that binds to a user-specified name a `GlobalId` produced by the Storage layer in response to a `Create` command.
+A created source has an initial read capability and write capability, which may not be at the beginning of time.
+Two sources created with the same arguments are not guaranteed to have the same contents at the same time (said differently: Materialize uses *nominal* rather than *structural* equivalence for sources).
+
+Sources remain active until a `DROP SOURCE` command is received, at which point the Adapter layer drops its read capability.
+
+One common specialization of source is the "table".
+The `CREATE TABLE` command introduces a new source that is not automatically populated by an external source, and is instead populated by `UpdateAndDowngrade` commands.
+The Adapter layer may use a write-ahead log to durably coalesce multiple writes at the same timestamp, as the `UpdateAndDowngrade` command does not otherwise allow this.
+The `DROP TABLE` command drops both the read and write capabilities for the source.
+
+## Indexes
+
+Materialize supports a `CREATE INDEX` command that results in a dataflow that computes and maintains the contents of the index's subject, in indexed representation.
+The `CREATE INDEX` command can be applied to any read-only view whose definition avoids certain non-deterministic functions (e.g. `now()`, `rand()`, environment variables).
+
+Materialize's main feature is that for any indexable view, one receives identical results whether re-evaluating the view from scratch or reading it from a created index.
+
+The `CREATE INDEX` command results in a `MaintainView` command for the Compute layer.
+The inputs to that command are `GlobalId` identifiers naming Storage or Compute collections.
+The command is issued with read capabilities all advanced to a common frontier (often: the maximum frontier of read capabilities in the input).
+The command returns a read capability for the indexed collection, which the Adapter layer maintains and perhaps regularly downgrades, as is its wont.
+The dataflow remains active until a corresponding `DROP INDEX` command is received by the Adapter layer, which discards its read capability for the indexed collection.
+
+## Transactions
+
+The Adapter layer provides interactive transactions through the `BEGIN`, `COMMIT`, and `ROLLBACK` commands.
+There are various restrictions on these transactions, primarily because the Storage and Compute layers require you to durably write before you can read from them.
+This makes read-after-write transactions challenging.
+Write-after-read transactions are implemented by **exclusively** performing all reads at the current timestamp and then writing at the next timestamp.
+Other read-only or write-only transactions may co-occur with a write-after-read transaction, but only one write-after-read transaction may span each transition from one time to the next.
```

Comment:
write-only transactions may not co-occur with write-after-read transactions, because the write-only txn could change the results of the write-after-read transaction.

Reply:
Ah, I meant to imply "the write half of the write-after-read transaction". Is there a better way to say that?

Reply:
Actually, while I did mean that, can you explain why it wouldn't be appropriate for a write to occur at the same timestamp as the read of a write-after-read transaction? The read would observe the write, and so I think the write-after-read transaction would be valid. I have to admit to not having thought deeply about this though.

Reply:
If the timestamps are equal it is ok. I think this was me understanding "co-occur" to mean "any timestamp between the write and read timestamp" and not "exactly at the read timestamp", which it sounds like is the actual definition of co-occur.

Reply:
I think this maybe comes down to me only thinking of the transactions as incrementing by one, where you are thinking about broader ranges of times. I'll try and make it all more clear!

Reply:
Text improved to spell out which co-occurrences are allowed.